pkg/syncer/resourcesync/controller.go

```diff
@@ -365,79 +330,163 @@ func (c *Controller) checkSSAR(ctx context.Context, gvr schema.GroupVersionResou
 	return sar.Status.Allowed, nil
 }
 
-// stopUnusedSyncerInformers stop syncers for gvrs not in requiredGVRs.
-func (c *Controller) stopUnusedSyncerInformers(ctx context.Context, requiredGVRs map[schema.GroupVersionResource]bool) {
+// removeUnusedGVRs stop syncers for gvrs not in requiredGVRs.
+func (c *Controller) removeUnusedGVRs(ctx context.Context, requiredGVRs map[schema.GroupVersionResource]bool) error {
 	logger := klog.FromContext(ctx)
 
 	c.mutex.Lock()
 	defer c.mutex.Unlock()
 
-	for gvr, informer := range c.syncerInformerMap {
+	updated := false
+	for gvr := range c.gvrs {
 		if _, ok := requiredGVRs[gvr]; !ok {
 			logger.WithValues("gvr", gvr.String()).V(2).Info("Stop syncer for gvr")
-			informer.cancel()
-			delete(c.syncerInformerMap, gvr)
+			delete(c.gvrs, gvr)
+			updated = true
 		}
 	}
+
+	if updated {
+		c.notifySubscribers(ctx)
+	}
+	return nil
 }
 
-func (c *Controller) startSyncerInformer(ctx context.Context, gvr schema.GroupVersionResource, syncTarget *workloadv1alpha1.SyncTarget) {
+func (c *Controller) GVRs() map[schema.GroupVersionResource]informer.GVRPartialMetadata {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+
+	gvrs := make(map[schema.GroupVersionResource]informer.GVRPartialMetadata, len(c.gvrs)+len(builtinGVRs)+1)
+	gvrs[corev1.SchemeGroupVersion.WithResource("namespaces")] = informer.GVRPartialMetadata{
+		Scope: apiextensionsv1.ClusterScoped,
+		Names: apiextensionsv1.CustomResourceDefinitionNames{
+			Singular: "namespace",
+			Kind:     "Namespace",
+		},
+	}
+	for key, value := range builtinGVRs {
+		gvrs[key] = value
+	}
+	for key, value := range c.gvrs {
+		gvrs[key] = value
+	}
+	return gvrs
+}
+
+func (c *Controller) getGVRPartialMetadata(gvr schema.GroupVersionResource) (*informer.GVRPartialMetadata, error) {
+	apiResourceList, err := c.cachedDiscovery.ServerResourcesForGroupVersion(gvr.GroupVersion().String())
+	if err != nil {
+		return nil, err
+	}
+	for _, apiResource := range apiResourceList.APIResources {
+		if apiResource.Name == gvr.Resource {
+			var resourceScope apiextensionsv1.ResourceScope
+			if apiResource.Namespaced {
+				resourceScope = apiextensionsv1.NamespaceScoped
+			} else {
+				resourceScope = apiextensionsv1.ClusterScoped
+			}
+
+			return &informer.GVRPartialMetadata{
+					Names: apiextensionsv1.CustomResourceDefinitionNames{
+						Kind:     apiResource.Kind,
+						Singular: apiResource.SingularName,
+					},
+					Scope: resourceScope,
+				},
+				nil
+		}
+	}
+	return nil, fmt.Errorf("unable to retrieve discovery for GVR: %s", gvr)
+}
+
+func (c *Controller) Ready() bool {
+	return c.synctargetInformerCacheSync()
+}
+
+func (c *Controller) Subscribe() <-chan struct{} {
```

Comment:
Subscribe to what?

Reply:
Added GoDoc that should make it clear. This is the implementation of the `informer.GVRSource.Subscribe()` method.